spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -444,85 +549,156 @@ public void run() {
 		}
 
 		private void invokeListener(final ConsumerRecord<K, V> record) {
+			final ConsumerAcknowledgment acknowledgment = this.acknowledgingMessageListener != null
+					? new ConsumerAcknowledgment(record) : null;
+			ListenableFuture<Void> future = getListenerTaskExecutor().submitListenable(new Callable<Void>() {
+
+				@Override
+				public Void call() throws Exception {
+					actualInvokeListener(record, acknowledgment);
+					return null;
+				}
+
+			});
 			try {
-				if (this.acknowledgingMessageListener != null) {
-					this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() { //NOSONAR - long anon.
+				future.get(this.pauseAfter, TimeUnit.MILLISECONDS);
+				if (this.logger.isDebugEnabled()) {
+					this.logger.debug("Success for " + record);
+				}
+				handleManualAcks();
+				this.listenerHasReturned = true;
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				throw new KafkaException("Listener thread interrupted", e);
+			}
+			catch (ExecutionException e) {
+				this.logger.debug("Failure for " + record + ": " + e.getMessage());
+				if (this.pauseWhenSlow && this.pauseException != null
+						&& this.pauseException.isAssignableFrom(e.getCause().getClass())) {
+					this.shouldPause = true;
+					this.listenerHasReturned = true;
+					this.pausedForException = true;
+				}
+				else {
+					if (getErrorHandler() != null) {
+						getErrorHandler().handle(e, record);
+					}
+					else {
+						this.logger.error("Listener threw an exception and no error handler for " + record, e);
+					}
+				}
+			}
+			catch (TimeoutException e) {
+				if (this.pauseWhenSlow) {
+					this.shouldPause = true;
+					future.addCallback(new ListenableFutureCallback<Void>() {
```

Comment:
Ah! Got it!
This is the main trick how to `resume` from the slow listener.
Since we don't `cancel` the `Future` we still can get the result and we do that via this callback, meanwhile the main loop dance with the void heartbeats.
Doh...